### PR TITLE
A couple of fixes & tweaks

### PR DESCRIPTION
**Pull request by NiLuJe**
* TextBoxWidget stuff for #8241
* Runtime selection of online CPU core count on Kobo (Elipsa) to allow spending most of our time with only a single core online.
  (The 2 core switch is actually probably mostly overkill still, but, oh, well ;)).
* BookInfoManager: Avoid undefined behavior in `collectSubprocesses` .

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8250)

### CODE REVIEW

File: frontend/ui/widget/textboxwidget.lua [hunk 1]
```diff
@@ -1088,7 +1088,12 @@ function TextBoxWidget:getCharPos()
 end
 
 function TextBoxWidget:getSize()
-    return Geom:new{ w = self.width, h = self._bb:getHeight()}
+    -- Make sure we actually have a BB, in case we're recycling an instance... (c.f., #8241)
+    if not self._bb then
+        self:_renderText(self.virtual_line_num, self.virtual_line_num + self.lines_per_page - 1)
```
Comment (Frenzie) [inline frontend/ui/widget/textboxwidget.lua:1093]: That takes care of creating one all by itself properly?

File: frontend/device/kobo/device.lua [hunk 2]
```diff
@@ -892,7 +898,53 @@ function Kobo:toggleChargingLED(toggle)
...
+local function getCPUCount()
+    local fd = io.open("/sys/devices/system/cpu/possible", "r")
+    if fd then
+        local str = fd:read("*line")
+        fd:close()
+
+        -- Format is n-N, where n is the first core, and N the last (e.g., 0-3)
+        return tonumber(str:match("%d+$")) or 0
+    else
+        return 0
```
Comment (Frenzie) [inline frontend/device/kobo/device.lua:914]: Wouldn't there logically be at least 1?

File: frontend/device/kobo/device.lua [hunk 3]
```diff
@@ -892,7 +898,53 @@ function Kobo:toggleChargingLED(toggle)
...
+    if not self:isSMP() then
+        return
+    end
+
+    if not self.cpu_count then
+        self.cpu_count = getCPUCount()
+    end
+
+    -- Not actually SMP or getCPUCount failed...
+    if self.cpu_count == 0 then
```
Comment (Frenzie) [inline frontend/device/kobo/device.lua:928]: Oh, never mind. :-P

File: frontend/document/koptinterface.lua [hunk 4]
```diff
@@ -381,6 +390,10 @@ function KoptInterface:renderOptimizedPage(doc, pageno, rect, zoom, rotation, re
 
     local cached = DocCache:check(hash, TileCacheItem)
     if not cached then
+        if hinting then
+            Device:enableCPUCores(2)
```
Comment (Frenzie) [inline frontend/document/koptinterface.lua:394]: Is that all that's actually useful in theory or just all that's actually useful in current practice?

File: frontend/ui/widget/textboxwidget.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/ui/widget/textboxwidget.lua:1093]: Yup!

File: frontend/document/koptinterface.lua (hunk 4 above)
Comment (NiLuJe) [reply frontend/document/koptinterface.lua:394]: It's already barely useful here; as we're heavily single-threaded ;p.

It's mainly for the sake of the eInk driver itself, which is fairly terrible (read: resource intensive ;p).

File: frontend/ui/widget/textboxwidget.lua (hunk 1 above)
Comment (poire-z) [reply frontend/ui/widget/textboxwidget.lua:1093]: https://github.com/koreader/koreader/issues/8241#issuecomment-926017666

> I'd prefer a simpler self:update(), to have it look less scary :)

Comment (NiLuJe) [reply frontend/ui/widget/textboxwidget.lua:1093]: And I find the extra layer of indirection this would add *more* confusing ;).

Comment (NiLuJe) [reply frontend/ui/widget/textboxwidget.lua:1093]: (And the `free` call inlined in `update` may come back to bite us in the ass, too).

Comment (poire-z) [reply frontend/ui/widget/textboxwidget.lua:1093]: Well, ok then.

(I would then probably wrap all others `self:_renderText(...)` in a `self:_update()` or something like that.)

Comment (NiLuJe) [reply frontend/ui/widget/textboxwidget.lua:1093]: > (I would then probably wrap all others `self:_renderText(...)` in a `self:_update()` or something like that.)

Yeah, that sounds better ;).

File: frontend/ui/widget/textboxwidget.lua [hunk 5]
```diff
@@ -860,6 +860,13 @@ function TextBoxWidget:_renderText(start_row_idx, end_row_idx)
     self:_renderImage(start_row_idx)
 end
 
+-- Lay out the full text, starting at the current line.
+-- (NOTE: This instantiates the inner bb (self._bb), so be careful about its lifecycle when you call this,
+--        c.f., TextBoxWidget:update).
+function TextBoxWidget:_updateLayout()
```
Comment (NiLuJe) [inline frontend/ui/widget/textboxwidget.lua:866]: Possibly `_updateTextLayout`, as it's not exactly the complete layout, @poire-z?

(Although, given the widget's name, text *is* most of the layout involved ;p).

Comment (poire-z) [reply frontend/ui/widget/textboxwidget.lua:866]: Either names are fine by me, as you prefer :)
(Althought the shorter _updateLayout() is just fine :)

**Review by poire-z — APPROVED** [review]

**Comment by NiLuJe** [comment]
(Now includes UB cleanup in `BookInfoManager:collectSubprocesses`, because I saw weird things when testing this ;p.

File: plugins/coverbrowser.koplugin/bookinfomanager.lua [hunk 6]
```diff
@@ -617,25 +617,27 @@ end
...
+        -- In-place removal, hence the reverse iteration.
+        for i = #self.subprocesses_pids, 1, -1 do
             local pid = self.subprocesses_pids[i]
             if FFIUtil.isSubProcessDone(pid) then
                 table.remove(self.subprocesses_pids, i)
                 -- Prevent has been issued for each bg task spawn, we must allow for each death too.
                 UIManager:allowStandby()
-            else
-                i = i + 1
             end
```
Comment (poire-z) [inline plugins/coverbrowser.koplugin/bookinfomanager.lua:633]: > Can't table.remove a table you're currently iterating on in order :s.

I don't think there was any issue: it's not like we were using ipairs, which would be messy with table.remove. Here, we were just iterating in ascending order by using an index and keepint it unchanged when table.remove'ing it.

I'm fine with your change, which makes it more readable - but I really think it fixes nothing.

File: plugins/coverbrowser.koplugin/bookinfomanager.lua [hunk 7]
```diff
@@ -722,8 +725,10 @@ function BookInfoManager:extractInBackground(files)
     -- and fill linux processes table)
     -- We set a single scheduled action for that
     if not self.subprocesses_collector then -- there's not one already scheduled
-        self.subprocesses_collector = UIManager:scheduleIn(
-            self.subprocesses_collect_interval, function()
+        self.subprocesses_collector = true
+        UIManager:scheduleIn(
```
Comment (poire-z) [inline plugins/coverbrowser.koplugin/bookinfomanager.lua:729]: > UIManager:scheduleIn doesn't return anything, using that as a boolean trap made no sense.

Odd, but it seems true, even in 2017 it returned nothing :/
So, ok :)

**Comment by poire-z** [comment]
> Now includes UB cleanup 

"UB" = undefined behaviour ? 
It feels it wasn't undefined, but quite defined: we allowed having multiple collectSubprocesses at the same time :)
Or is it something else?

File: plugins/coverbrowser.koplugin/bookinfomanager.lua (hunk 6 above)
Comment (NiLuJe) [reply plugins/coverbrowser.koplugin/bookinfomanager.lua:633]: True ;o).

File: frontend/ui/widget/textboxwidget.lua [hunk 8]
```diff
@@ -1122,6 +1134,7 @@ function TextBoxWidget:free(full)
         self.cursor_restore_bb:free()
         self.cursor_restore_bb = nil
     end
+    self.line_num_to_image = nil
```
Comment (poire-z) [inline frontend/ui/widget/textboxwidget.lua:1137]: Bad idea: thumbnails in Wikipedia results and full articles are not shown after this (the room for the image is still there, but not image displayed, and long press in the hold select texts around).
Remove this line and all works again.
(Except for the numerous segfault I get when existing KOreader :/)

**Pull request merged**